### PR TITLE
Bybit: fetchOpenInterest

### PR DESCRIPTION
**Pull request by Dan-krm**
Added a new unified method `fetchOpenInterest` to Bybit:
```
node examples/js/cli bybit fetchOpenInterest BTC/USD:USDC

bybit.fetchOpenInterest (BTC/USD:USDC)
2022-09-28T23:16:28.484Z iteration 0 passed in 299 ms

{
  symbol: 'BTC/USD:USDC',
  openInterestAmount: 44142800000,
  openInterestValue: undefined,
  timestamp: 1664406000000,
  datetime: '2022-09-28T23:00:00.000Z',
  info: {
    symbol: 'BTCPERP',
    timestamp: '1664406000',
    openInterest: '44142800000'
  }
}
2022-09-28T23:16:28.484Z iteration 1 passed in 299 ms
```
```
node examples/js/cli bybit fetchOpenInterest BTC/USDT:USDT

bybit.fetchOpenInterest (BTC/USDT:USDT)
2022-09-28T23:18:08.972Z iteration 0 passed in 302 ms

{
  symbol: 'BTC/USDT:USDT',
  openInterestAmount: 55558.919,
  openInterestValue: undefined,
  timestamp: 1664406000000,
  datetime: '2022-09-28T23:00:00.000Z',
  info: {
    open_interest: '55558.919',
    timestamp: '1664406000',
    symbol: 'BTCUSDT'
  }
}
2022-09-28T23:18:08.972Z iteration 1 passed in 302 ms
```

### CODE REVIEW

File: doc/manual.rst [hunk 1]
```diff
@@ -4459,6 +4459,28 @@ Funding Rate History Structure
...
+
+
+ * **symbol** (String) Unified CCXT market symbol (e.g. ``"BTC/USDT:USDT"``\ )
+ * **params** (Dictionary) Extra parameters specific to the exchange API endpoint (e.g. ``{"endTime": 1645807945000}``\ )
+
+Returns
+
+
+ * A dictionary of :ref:`open interest structures <open interest structure>`
+
```
Comment (samgermain) [inline doc/manual.rst:4483]: I think manual.rst is a compiled file. When I add a section to the manual, I add it inside `wiki/Manual.md`

Also add the method name to https://github.com/ccxt/ccxt/blob/master/wiki/Manual.md#unified-api

File: doc/manual.rst [hunk 2]
```diff
@@ -4493,8 +4515,10 @@ Open Interest Structure
 
    {
        symbol: 'BTC/USDT',
-       baseVolume: 80872.801,
-       quoteVolume: 3508262107.38,
+       baseVolume: 80872.801,  // deprecated
+       quoteVolume: 3508262107.38, // deprecated
```
Comment (samgermain) [inline doc/manual.rst:4519]: If you're going to deprecate these fields, I think you should create another PR where the new fields are added for every exchange that has `parseOpenInterest`

File: js/bybit.js [hunk 3]
```diff
@@ -4859,7 +4862,12 @@ module.exports = class bybit extends Exchange {
         if (limit !== undefined) {
             request['limit'] = limit;
         }
-        const response = await this.publicGetV2PublicOpenInterest (this.extend (request, params));
+        const isUsdcSettled = market['settle'] === 'USDC';
+        let method = 'publicGetV2PublicOpenInterest';
+        if (isUsdcSettled) {
+            method = 'publicGetPerpetualUsdcOpenapiPublicV1OpenInterest';
+        }
+        const response = await this[method] (this.extend (request, params));
```
Comment (samgermain) [inline js/bybit.js:4870]: The edits to `fetchOpenInterestHistory` should be made in another PR

File: js/bybit.js [hunk 4]
```diff
@@ -4877,27 +4885,127 @@ module.exports = class bybit extends Exchange {
...
+         * @description Retrieves the open interest of a currency
+         * @see https://bybit-exchange.github.io/docs/futuresV2/inverse/#t-marketopeninterest
+         * @see https://bybit-exchange.github.io/docs/usdc/perpetual/#t-queryopeninterest
+         * @param {string} symbol Unified CCXT market symbol
+         * @param {object} params exchange specific parameters
+         * @param {string|undefined} params.timeframe "5m", 15m, 30m, 1h, 4h, 1d
+         * @param {int|undefined} params.limit The number of open interest structures to return. Max 200, default 50
+         * @returns {object} an open interest structure{@link https://docs.ccxt.com/en/latest/manual.html#interest-history-structure}
+         */
+        const timeframe = this.safeString (params, 'timeframe', '1h');
```
Comment (samgermain) [inline js/bybit.js:4919]: Should `timeframe` just be a parameter? Have you written this method for a few more exchanges and found out what the common parameters would be?

**Review by samgermain — CHANGES_REQUESTED** [review]

File: js/bybit.js (hunk 4 above)
Comment (Dan-krm) [reply js/bybit.js:4919]: Bybit is the only exchange I've seen use `timeframe` so far, usually they only require a market id and return the current open interest

File: wiki/Manual.md [hunk 5]
```diff
@@ -2848,6 +2849,25 @@ Returns
...
+
+Parameters
+
+- **symbol** (String) Unified CCXT market symbol (e.g. `"BTC/USDT:USDT"`)
+- **params** (Dictionary) Extra parameters specific to the exchange API endpoint (e.g. `{"endTime": 1645807945000}`)
+
+Returns
+
+- A dictionary of [open interest structures](#open-interest-structure)
+
```
Comment (samgermain) [inline wiki/Manual.md:2870]: Sorry, let's actually move this addition to a new PR, and let's implement `fetchOpenInterest` on several more exchanges before merging in the documentation, because it's get's more difficult to change the method after that

Comment (Dan-krm) [reply wiki/Manual.md:2870]: Okay no problem it's already implemented on okx, huobi, aax and ftx, there's also a bitget PR pending

Comment (Dan-krm) [reply wiki/Manual.md]: https://github.com/ccxt/ccxt/pull/15254

**Comment by carlosmiei** [comment]
@Dan-krm Can you please update this method with the v3 endpoints? Bybit will probably deprecate v2 soon

**Comment by Dan-krm** [comment]
> @Dan-krm Can you please update this method with the v3 endpoints? Bybit will probably deprecate v2 soon

Yeah definitely, I'll update the fetchOpenInterestHistory PR as well

File: js/bybit.js [hunk 6]
```diff
@@ -4881,23 +4882,78 @@ module.exports = class bybit extends Exchange {
...
+        if (timeframe === '1m') {
+            throw new BadRequest (this.id + ' fetchOpenInterest() cannot use the 1m timeframe');
+        }
+        await this.loadMarkets ();
+        let market = this.market (symbol);
+        if (!market['contract']) {
+            throw new BadRequest (this.id + ' fetchOpenInterest() supports contract markets only');
+        }
+        const defaultSubType = this.safeString (this.options, 'defaultSubType', 'linear');
+        const category = this.safeString (params, 'category', defaultSubType);
```
Comment (carlosmiei) [inline js/bybit.js:4907]: @Dan-krm do we need to re-work our symbols? otherwise why wouldn't we read the subType from the symbol itself?

Comment (Dan-krm) [reply js/bybit.js:4908]: I've added this line `const linear = market['linear'] ? 'linear' : 'inverse';` let me know if that's not what you had in mind

Comment (carlosmiei) [reply js/bybit.js:4908]: @Dan-krm that's exactly what I had in mind, but apparently, it's not working for inverse markets: 
image

Comment (Dan-krm) [reply js/bybit.js:4908]: @carlosmiei I'm receiving this error for both linear and inverse now:
```
[InvalidNonce] bybit {"ret_code":10002,"ret_msg":"invalid request,please check your server timestamp or recv_window param","ext_code":"","result":null,"ext_info":null,"time_now":1666068788802}

    at throwExactlyMatchedException  js/base/Exchange.js:2140              throw new exact[string] (message);                                                 
    at handleErrors                  js/bybit.js:5447                      this.throwExactlyMatchedException (this.exceptions['exact'], errorCode, feedbac…   
    at                               js/base/Exchange.js:598               const skipFurtherErrorHandling = this.handleErrors (response.status, response.s…   
    at processTicksAndRejections     node:internal/process/task_queues:96                                                                                     
    at async timeout                 js/base/functions/time.js:181         return await Promise.race ([promise, expires.then (() => { throw new TimedOut (…   
    at fetch2                        js/base/Exchange.js:1801              return await this.fetch (request['url'], request['method'], request['headers'],…   
    at request                       js/base/Exchange.js:1805              return await this.fetch2 (path, api, method, params, headers, body, config, con…   
    at fetchCurrencies               js/bybit.js:653                       const response = await this.privateGetAssetV1PrivateCoinInfoQuery (params);        
    at loadMarketsHelper             js/base/Exchange.js:624               currencies = await this.fetchCurrencies ()                                         
    at async run                     examples/js/cli.js:231                await exchange.loadMarkets ()                                                      

bybit {"ret_code":10002,"ret_msg":"invalid request,please check your server timestamp or recv_window param","ext_code":"","result":null,"ext_info":null,"time_now":1666068788802}
```

Comment (carlosmiei) [reply js/bybit.js:4908]: @Dan-krm but that's a different error, usually means a difference between your clock and the exchange.

Comment (Dan-krm) [reply js/bybit.js:4908]: @carlosmiei with the current ordering defaultSubType in the exchange options need to be set to inverse for it to work properly, I think removing defaultSubType from fetchOpenInterest is the best approach

File: js/bybit.js [hunk 7]
```diff
@@ -4881,23 +4882,78 @@ module.exports = class bybit extends Exchange {
...
+        //        "openInterest": 64757.62400000,
+        //        "timestamp": 1665784800000,
         //    }
         //
-        const id = this.safeString (interest, 'symbol');
-        market = this.safeMarket (id, market);
-        const timestamp = this.safeTimestamp (interest, 'timestamp');
-        const numContracts = this.safeString (interest, 'open_interest');
+        const timestamp = this.safeInteger (interest, 'timestamp');
+        const numContracts = this.safeString (interest, 'openInterest');
```
Comment (carlosmiei) [inline js/bybit.js:4948]: we have to keep supporting `open_interest` otherwise will break `fetchOpenInterestHistory` parsing

File: js/bybit.js [hunk 8]
```diff
@@ -4881,23 +4882,78 @@ module.exports = class bybit extends Exchange {
...
         const contractSize = this.safeString (market, 'contractSize');
+        const value = this.parseNumber (Precise.stringMul (numContracts, contractSize));
```
Comment (carlosmiei) [inline js/bybit.js:4950]: Since the contract size is 1, I was wondering if this makes sense because we're saying openInterestAmount = openInterestValue,  what do you think?

File: js/bybit.js (hunk 7 above)
Comment (Dan-krm) [reply js/bybit.js:4948]: Okay I can add it back, I use a different endpoint for `fetchOpenInterestHistory` in this other PR that no longer uses `open_interest` https://github.com/ccxt/ccxt/pull/15088

File: js/bybit.js (hunk 8 above)
Comment (Dan-krm) [reply js/bybit.js:4950]: I agree I don't think it makes sense, the value for numContracts is the amount of open interest in BTC so I think that variable name isn't accurate as well

**Comment by carlosmiei** [comment]
LGTM ✅ 

```
 p bybit fetchOpenInterest "BTC/USDT:USDT" --sandbox
Python v3.9.7
CCXT v1.93.108
bybit.fetchOpenInterest(BTC/USDT:USDT)
{'baseVolume': 85727.73299999,
 'datetime': '2022-10-24T09:00:00.000Z',
 'info': {'openInterest': '85727.73299999', 'timestamp': '1666602000000'},
 'openInterestAmount': None,
 'openInterestValue': 85727.73299999,
 'quoteVolume': None,
 'symbol': 'BTC/USDT:USDT',
 'timestamp': 1666602000000}
```

**Review by carlosmiei — APPROVED** [review]
